### ISSUE TITLE
Mocking external functions in R6 classes when using expressions

### ISSUE THREAD

**Issue by aalucaci** [r-lib/mockery#40]
Hello,
in view of the 3rd edition of the testthat package, I currently try to see if I can replace `testthat::with_mock` with `mockery::stub`. I have the problem that I cannot mock external functions, which are called via expressions in a method of an R6 class.

The detailed example is given here:

[https://github.com/r-lib/testthat/issues/1185](https://github.com/r-lib/testthat/issues/1185)

Is there any chance that this can be implemented in the mockery package?

Kind regards,
Angela

**Comment by nbenn** [comment]
@aalucaci I was also having issues with switching from `testthat::with_mock()` to `mockery::stub()` which ultimately led me to start working on [my own solution](https://github.com/nbenn/mockthat/). Using the current development version of `mockthat::with_mock()` should do what you want to:

```r
obj <- A$new()
#> initialize

obj$run1()
#> this is the external_function

mockthat::with_mock(external_function = "this is the mock", obj$run1())
#> this is the mock
```

This should be available from CRAN shortly, in case you're interested.

**Comment by aalucaci** [comment]
@nbenn That's very good news, I kind of gave up hope that I'll ever receive an answer. 

I tested your example and it works indeed as you said. Do you know when will your package be available on CRAN?

Kind regards,
Angela

**Comment by nbenn** [comment]
@aalucaci v0.2.4 is now available for download from CRAN. I have already fixed a couple of things as v0.2.5 and I will submit that update to CRAN in ~2 weeks as to not overload the CRAN team. If you run into issue I'm happy to take them over at https://github.com/nbenn/mockthat/issues.

**Comment by aalucaci** [comment]
@nbenn thank you for your reply. I'll use your package and come back to you in case of issues.

Closing now this issue, since thanks to you now I have a solution to my problem.

Thank you once again,
Angela
